aten/src/ATen/native/cuda/Loops.cuh

```diff
@@ -131,6 +139,164 @@ invoke(const func_t &f, char *const C10_RESTRICT data[], const index_t strides[]
+}
+
+template <typename func_t, typename index_t, typename traits = function_traits<func_t>>
+C10_HOST_DEVICE typename traits::result_type
+invoke_with_index(const func_t &f, char* const C10_RESTRICT data[], const index_t strides[], const ScalarType dtypes[],
+                  int i, int idx) {
+  // index at last position
+  using Indices = std::make_index_sequence<traits::arity-1>;
+  return invoke_with_index_impl<traits>(f, data, strides, dtypes, i, idx, Indices{});
+}
+
+} // namespace legacy
+
+// See the note for namespace legacy above.
+namespace modern {
+
+namespace detail {
+
+template <typename func_t, typename array_t, std::size_t... I>
+__device__ inline constexpr decltype(auto) apply_impl(func_t f, array_t t, std::index_sequence<I...>)
+{
+    return f(t[I]...);
+}
+
+template <typename func_t, typename array_t>
+__device__ inline constexpr decltype(auto) array_apply(func_t f, array_t a) {
+  constexpr std::size_t arity = function_traits<func_t>::arity;
+  return apply_impl(f, a, std::make_index_sequence<arity>{});
+}
+
+}  // namespace detail
+
+template<int nt, int vt, typename func_t, typename array_t, std::enable_if_t<(function_traits<func_t>::arity > 0), int> = 0>
+C10_LAUNCH_BOUNDS_1(nt)
+__global__ void elementwise_kernel(int N, func_t f, array_t data) {
+  // Assumption:
+  // all arguments have the same type, which could be different from the return type
+  // the all tensors are contiguous, that is: stride == sizeof(type) for all tensors
+
+  using traits = function_traits<func_t>;
+  using return_t = typename traits::result_type;
+  using arg_t = typename traits::template arg<0>::type;
+  constexpr int arity = traits::arity;
+
+  int tid = threadIdx.x;
+  int nv = nt * vt;
+  int idx = nv * blockIdx.x + tid;
+
+  // compute base pointers
+  return_t *result_base = reinterpret_cast<return_t *>(data[0]) + idx;
+  arg_t *args_base[arity];
+  #pragma unroll
+  for (int i = 0; i < arity; i++) {
+    args_base[i] = reinterpret_cast<arg_t *>(data[i + 1]) + idx;
+  }
+
+  // fetch data
+  return_t results[vt];
+  arg_t args[vt][arity];
+  #pragma unroll
+  for (int i = 0; i < vt; i++) {
+    if (idx + nt * i < N) {
+      #pragma unroll
+      for (int j = 0; j < arity; j++) {
+        args[i][j] = *(args_base[j] + i * nt);
+      }
+    }
+  }
+
+  // compute
+  #pragma unroll
+  for (int i = 0; i < vt; i++) {
+    if (idx + nt * i < N) {
+      results[i] = detail::array_apply(f, args[i]);
+    }
+  }
+
+  // store data
+  #pragma unroll
+  for (int i = 0; i < vt; i++) {
+    if (idx + nt * i < N) {
+      *(result_base + i * nt) = results[i];
+    }
+  }
+}
+
+template<int nt, int vt, typename func_t, typename array_t, std::enable_if_t<function_traits<func_t>::arity == 0, int> = 0>
+C10_LAUNCH_BOUNDS_1(nt)
+__global__ void elementwise_kernel(int N, func_t f, array_t data) {
+  // Assumption:
+  // all arguments have the same type, which could be different from the return type
```

Comment:
this comment is misleading, there are no arguments?

Reply:
arguments for `f`. I will modify the comment